docs: add overview of session recording (#18717)
gravitational/teleport: docs/pages/reference/session-recording.mdx

---
title: Teleport Session Recording
description: An overview of Teleport's session recording and its configuration
---

This page provides an overview of how Teleport records sessions and the various
configuration options that apply to session recordings.

Teleport supports two different recording configurations: synchronous recording
and asynchronous recording. Additionally, Teleport administrators can configure
where the recording takes place: at the node or at the proxy.

This results in 4 possible sesssion recording configurations:

- `node-sync`: synchronous recording performed by the SSH node

> - `node`: asynchronous recording performed by the SSH node

Author: > We're replacing "SSH" with "Server Access", right?

That would be news to me. Though I don't see a harm in sticking with SSH here since it's the most specific.

> This is true for Server Access sessions, but what about others?

Well, that's the thing, SSH is unique in that Teleport software is running at both ends of the connection (Teleport is the SSH server you're connecting to, and Teleport is the SSH client)

For desktop access (as mentioned elsewhere in this PR) there is not a distinction between "node" and "proxy" - the recording is always done on the windows_desktop_service, which is neither a node or proxy. 

Kube and DB access are similar - the Kube cluster or DB that you're connecting to is not Teleport software, it's other software that's totally unaware of Teleport, just like a desktop with RDP.

> - `node`: asynchronous recording performed by the SSH node

Reviewer: > > We're replacing "SSH" with "Server Access", right?
> 
> That would be news to me. Though I don't see a harm in sticking with SSH here since it's the most specific.

Maybe I have that backwards.

> - `node`: asynchronous recording performed by the SSH node

Reviewer: > Kube and DB access are similar - the Kube cluster or DB that you're connecting to is not Teleport software, it's other software that's totally unaware of Teleport, just like a desktop with RDP.

You mean similar to Desktop, where the recoding is done at the node running the service, right?

> - `node`: asynchronous recording performed by the SSH node

Reviewer: So the more accurate statement is that session recording for DB and Kube occurs at the host running the Teleport Service for Database and Kubernetes access?

> - `node`: asynchronous recording performed by the SSH node

Author: Right. Kube, DB, Desktop (and even App Access) are all alike in this regard, while SSH is a bit unique, which is why I mentioned SSH specifically in the original bullet point that started this thread.

- `proxy-sync`: synchronous recording performed at the proxy
- `proxy`: asynchronous recording performed at the proxy

This is a cluster-wide configuration option and applies to the entire
Teleport cluster. If you need to apply different recording configuration to
different sets of resources, we recommend setting up Trusted Clusters with
their own recording configurations.

<Admonition type="note" title="Desktop Access">

> Desktop Sessions are always recorded at Teleport's Windows Desktop Service
> since there is no Teleport software running on the "node" (desktop).
> 
> For this reason, Teleport Desktop Access treats `node` and `proxy` identically
> (perform synchronous desktop recording) and `node-sync` and `proxy-sync` identically
> (perform asynchronous desktop recording).

Reviewer: Suggested to address the discussion thread above.

```suggestion
Desktop, Database, and Kubernetes Sessions are always recorded at the host running
the Teleport service for that session type, since there is no Teleport software
running on the "node" (desktop, database, or Kubernetes cluster).

For this reason, Teleport Desktop, Database, and Kubernetes Access treats `node`
and `proxy` identically (perform synchronous recording) and `node-sync` and
`proxy-sync` identically (perform asynchronous recording).
```

Reviewer: Adjusted after discussion.

specified by the `audit_sessions_uri` field in the `teleport.yaml` configuration file.
Teleport current supports the following session storage backends:

- File: stores recordings on the local filesystem. Suitable for dev environments, demos,
  and small home environments.
- S3: stores recordings in an AWS S3 bucket, or S3-compatible storage.
  Suitable for production deployments.
- GCS: stores recordings in Google Cloud Storage. Suitable for production deployments.

The Teleport Auth server is the only component that writes directly to the storage backend.

Note that the session storage backend is distinct from Teleport's cluster state or
audit log storage, which supports a different set of services (SQLite, DynamoDB, Firestore, etc)

## Format

A Teleport session recording is an ordered sequence of structured events associated
with a session. Each event is an encoded [Protocol Buffer](https://developers.google.com/protocol-buffers)
and the complete session is compressed with gzip before being written to the storage backend.

These recordings have a `.tar` extension for backwards compatibility reason, but it

> should be noted that they are not TAR archives and cannot be read using the `tar` utility.

Reviewer 2: Someone told me once that it stands for **T**eleport **Ar**chive 😁

---
title: Teleport Session Recording
description: An overview of Teleport's session recording and its configuration
---

> This page provides an overview of how Teleport records sessions and the various

Reviewer 3: We might want to expand on what session recording is earlier in the page. What is a session? What does it mean to record it?

Author: Perhaps I'm blinded by my own experience, as it's pretty obvious to me. Do you have any suggestions for what we might say in an expanded explanation?

configuration options that apply to session recordings.

> Teleport supports two different recording configurations: synchronous and

Author: Ack. I'm being very careful to say "configurations" and not "recording modes" here, because unfortunately some recent work has stolen the term "recording modes" and I don't want to go overboard in this doc.

asynchronous recording. Additionally, Teleport administrators can configure

> where the recording takes place: at the node or at the proxy.

Reviewer 3: We should make "node" and "proxy" are consistent with the rest of the docs (in terms of usage and capitalization), especially the Core Concepts page (https://goteleport.com/docs/core-concepts/). Does "proxy" refer to "Proxy Service" here? Also, does "node" refer to an SSH Service instance?

Author: Kind of. The use of these terms is somewhat intentional, as:

1. the config setting for these are literally `node` or `proxy`.
2. "proxy" typically means the "Proxy Service" but may also refer to the kube service or the windows desktop service, so the more generic proxy makes sense.
3. yes, 'node" refers to the SSH service

This results in 4 possible sesssion recording configurations:

- `node-sync`: synchronous recording performed by the SSH node
- `node`: asynchronous recording performed by the SSH node
- `proxy-sync`: synchronous recording performed at the proxy
- `proxy`: asynchronous recording performed at the proxy

This is a cluster-wide configuration option and applies to the entire
Teleport cluster. If you need to apply different recording configuration to
different sets of resources, we recommend setting up [Trusted Clusters](/docs/management/admin/trustedclusters/)
with their own recording configurations.

<Admonition type="note">

> Windows, Database, and Kubernetes Sessions are always recorded at the host running

Reviewer 3: ```suggestion
Windows, Database, and Kubernetes sessions are always recorded at the host running
```

> Windows, Database, and Kubernetes sessions are always recorded at the host running

Author: What this is trying to say is that the node/proxy distinction only matters for SSH sessions. For desktop or kubernetes sessions, `node` and `proxy` are equivalent, and `node-sync` and `proxy-sync` are equivalent.

If you have suggestions for a better way to phrase this I'm all ears!

the Teleport service for this session type, since there is no Teleport software
running on the "node" (desktop, database, or Kubernetes cluster).

For this reason, Teleport Windows, Database, and Kubernetes Access treats `node`
and `proxy` identically (perform synchronous recording) and `node-sync` and
`proxy-sync` identically (perform asynchronous recording).
</Admonition>

## Record at Node

By default, Teleport performs recording at the SSH node. This is because Teleport's
Proxy Server cannot see the SSH traffic to the node, it is encrypted end-to-end

> (from the SSH client to the SSH server).

Author: SSH client can be `tsh` or OpenSSH's `ssh` client.

SSH server can be Teleport's SSH service or OpenSSH's sshd.

that the system has enough disk space to accommodate the expected number of Teleport
sessions. Additionally, since recording data is temporarily stored on disk, there
is a greater chance that it can be tampered with, deleted, or otherwise corrupted
before the upload completes.

The advantage of asynchronous recording is that it doesn't require a persistent
connection to the Auth Server. For example, an SSH session can continue to operate
even if Teleport's Auth Server goes down. When the session completes Teleport will
attempt to upload the recording to the Auth Server. If the Auth Server is still
unavailable, Teleport has built-in retry and backoff mechanisms that will upload
the artifact when the Auth Server comes back online. Additionally, asynchronous
recording is well-suited towards recording sessions that are extra chatty or in
environments where the connection to the auth server is unreliable or high-latency.

## Storage

Session recordings are stored in Teleport's audit sessions backend, which is
specified by the `audit_sessions_uri` field in the `teleport.yaml` [configuration file](config.mdx).

> Teleport current supports the following session storage backends:

Reviewer 3: Should we list the values you'd put in the configuration file below?

Author: Personally, I don't think so. There's quite a few permutations and it's already covered in detail in the config reference. This page was more meant to be about how it works rather than how to configure it.

- File: stores recordings on the local filesystem. Suitable for dev environments, demos,
  and small home environments.
- S3: stores recordings in an AWS S3 bucket, or S3-compatible storage.
  Suitable for production deployments.
- GCS: stores recordings in Google Cloud Storage. Suitable for production deployments.

> The Teleport Auth server is the only component that writes directly to the storage backend.

Reviewer 3: ```suggestion
The Teleport Auth Service is the only component that writes directly to the storage backend.
```

Note that the session storage backend is distinct from Teleport's cluster state or

> audit log storage, which supports a different set of services (SQLite, DynamoDB, Firestore, etc)

Reviewer 3: ```suggestion
audit log storage, which supports a different set of services (SQLite, DynamoDB, Firestore, etc).
```

## Format

A Teleport session recording is an ordered sequence of structured events associated
with a session. Each event is an encoded [Protocol Buffer](https://developers.google.com/protocol-buffers)
and the complete session is compressed with gzip before being written to the storage backend.

These recordings have a `.tar` extension for backwards compatibility reasons, but it
should be noted that they are not TAR archives and cannot be read using the `tar` utility.

## Playback

SSH and Kubernetes sessions can be played in Teleport's Web UI or by using
the [`tsh play`](cli.mdx#tsh-play) command. Desktop session recordings can only be played back
in the Web UI.

In the Web UI, the session recordings page is populated by querying Teleport's
audit log for session end events, which contain metadata about the recording.
This is sometimes surprising to Teleport users, because even though the recordings
and audit log are stored in separate backends, both need to be operational in
order to play recordings.

> ## Upload Completer

Reviewer 3: ```suggestion
## Upload completer
```
Since we're lowercasing "upload completer"

Every Teleport process runs a service called the *upload completer* which periodically
checks for abandoned uploads and completes them if there is not an active session tracker
for the session associated with the recording. By default, the upload completer runs
every 5 minutes, and session trackers have a 30 minute expiration period. This means it
can take up to ~35 minutes after the service comes back online for an abandoned upload
to be completed.

In asynchronous recording modes, if the node goes down during the session, the partially
completed recording will sit on the node's disk. The node's upload completer will eventually
detect the abandoned upload and stream it to the Teleport Auth Server where it will be
written to the storage backend.

In synchronous recording modes, Teleport's Auth Server is streaming the recording directly
to storage. If the Auth Server goes down during a session, the uncompleted upload will sit
as a series of parts (in cloud storage or on the Auth Server's disk) and it is the
responsibility of the Auth Server's upload completer to detect the abandoned upload and
complete it.

> ## Related Reading

Reviewer 3: ```suggestion
## Related reading
```

---
title: Teleport Session Recording
description: An overview of Teleport's session recording and its configuration
---

This page provides an overview of how Teleport records sessions and the various

> configuration options that apply to session recordings. The content that is
> captured depends on the type of the session.

Reviewer 3: ```suggestion
configuration options that apply to session recordings. 

## What session recordings capture

The content session recordings capture depends on the type of the session.
```
Since this information is pretty granular, I thought it would make sense to separate it from the intro paragraph.

- SSH sessions: Teleport captures the entire pseudo-terminal (PTY) output
  of the session
- Kubernetes sessions: Teleport captures the entire PTY output for
  `kubectl exec` invocations.
- Desktop sessions: Teleport captures the contents of the desktop screen
  and any mouse input. Teleport does not capture keystrokes in the remote
  desktop.
- App sessions: Teleport captures a stream of `app.session.request`
  audit events related to application access. The audit events are
  bucketed into 5-minute time intervals informally referred to as "chunks."
- Database sessions: Teleport captures a stream of audit events
  related to the database being accessed.

> Teleport supports two different recording configurations: synchronous and

Reviewer 3: ```suggestion
## Session recording configurations

Teleport supports two different recording configurations: synchronous and
```
Suggested because this is a different topic area than the earlier breakdown of what session recording captures

- `node`: asynchronous recording performed by an SSH Service instance
- `proxy-sync`: synchronous recording performed on a Proxy Service instance
- `proxy`: asynchronous recording performed on a Proxy Service instance

This is a cluster-wide configuration option and applies to the entire
Teleport cluster. It can be configured in by setting `session_recording`
in the `auth_service` section of your `teleport.yaml`, or dynamically via
the the `session_recording_config` resource. If you need to apply different
recording configuration to different sets of resources, we recommend setting up
[Trusted Clusters](/docs/management/admin/trustedclusters/) with their own
recording configurations.

<Admonition type="note">
Windows, database, and Kubernetes sessions are always recorded at the host running
the Teleport service for this session type, since there is no Teleport software
running on the "node" (desktop, database, or Kubernetes cluster).

For this reason, Teleport Windows, Database, and Kubernetes Access treats `node`
and `proxy` identically (perform synchronous recording) and `node-sync` and
`proxy-sync` identically (perform asynchronous recording).
</Admonition>

> ## Record at Node

Reviewer 3: ```suggestion
### Record at Node
```
To place this in the "Session recording configurations" section I suggested earlier

By default, Teleport performs recording at the SSH node. This is because Teleport's
Proxy Server cannot see the SSH traffic to the node, it is encrypted end-to-end
(from the SSH client to the SSH server).

> ## Record at Proxy

Reviewer 3: ```suggestion
### Record at Proxy
```

In some cases, it is desirable to perform the recording at Teleport's Proxy Server.
For example, sessions to non-Teleport servers (like OpenSSH's `sshd`) can only be

> recorded at the proxy. This is referred to as *Recording Proxy Mode*.

Reviewer 3: ```suggestion
recorded at a Proxy Service instance. This is referred to as *Recording Proxy Mode*.
```

In this mode, the Proxy Server terminates (decrypts) the SSH connection and establishes
its own SSH connection to the destination server, effectively becoming an authorized
"man in the middle."

We consider this mode to be less secure, as it grants additional privileges to the
Proxy Server. Since the Proxy Server needs credentials to decrypt the SSH connection,
it must be properly secured and is a higher value target for an attacker than a proxy
that cannot decrypt the data flowing through it.

> Additionally, the credentials that the proxy uses to decrypt the SSH connection are

Reviewer 3: ```suggestion
Additionally, the credentials that the Proxy Server uses to decrypt the SSH connection are
```

provided via SSH Agent Forwarding, so Agent Forwarding must be enabled to record at
the Proxy Server.

> ## Synchronous recording

Reviewer 3: ```suggestion
### Synchronous recording
```

When synchronous recording is enabled, the Teleport component doing the recording

> (which may be the node or the proxy depending on your configuration) submits each

Reviewer 3: ```suggestion
(which may be the Node or the Proxy Service instance depending on your configuration) submits each
```

recording event to Teleport's Auth Server as it occurs. In this mode, failure to
emit a recording event is considered fatal - the session will be terminated if an
event cannot be recorded. This makes synchronous recording best suited for highly
regulated environments where you need to be confident that all data is recorded.
This also means that you need a reliable and low-latency connection to the Auth
Server for the duration of the session to ensure that the session isn't interrupted
or terminated due to temporary connection loss.

In synchronous recording modes, the Auth Server receives a stream of recording
events and is responsible for assembling them into the final artifact and uploading
it to the storage backend. Since data is streamed directly to the Auth Server,

> Teleport administrators don't need to be concerned with disk space on their nodes
> and proxies, as no recording data is written to those disks.

Reviewer 3: ```suggestion
Teleport administrators don't need to be concerned with disk space on their Nodes
and Proxy Service instances, as no recording data is written to those disks.
```

> ## Asynchronous recording

Reviewer 3: ```suggestion
### Asynchronous recording
```
